### DISCUSSION TITLE
New feature: Indicators

### DISCUSSION THREAD

**Discussion by pyscripter** [TurboPack/SynEdit#160] [Ideas]
This is inspired by [Scintilla indicators](https://www.scintilla.org/ScintillaDoc.html#Indicators).  Their most common usage is to indicate syntax errors or spelling mistakes by decorating the text.   Scintilla defines 20 such indicators, the most familiar of which is  the squiggly lines drawn to indicate errors.   This would simplify the implementation of  plugins such as the Spelling addin of SynEdit.

The suggestion is to implement this using events:

```pascal
TSynIndicator = record
  GUID: TGUID
  Style: TSynIndicatorStyle;
  Color: TColor;
  Alpha: Single;
end;

TSynIndicatorRange = record
  Indicator: TSynIndicator;
  BeginChar, EndChar: Integer;
end;

TSynIndicatorRangeArray = TArray<TSynRangeIndicator>;
// events
TSynGetIndicatorsForLine  = procedure(Sender: TObject; const Line: Integer;  IndicatorRanges: TSynIndicatorRangeArray;
// returns true if handled
TSynOnIdicatorHover = function(Sender; TObject; const MousePos; TPoint;  Line: Integer; IndicatorRange: TSynIndicatorRange): Boolean;
```

The first event will provide information about indicators to SynEdit text painting and the second event could give the opportunity to for example  show a hint or popup menu or take some other action when when the user hovers the mouse on an indicator range.
These events can be implemented on the SynEdit or plugin level,

What do you think?

**Comment by pyscripter** [comment]
## Implementation
The first implementation has been added to the main repo. This is the core of  the implementation in SynEditMiscClasses:

```pascal
  TSynIndicatorStyle = (sisTextDecoration, sisSquiggleMicrosoftWord,
    sisSquiggleWordPerfect, sisRectangle, sisFilledRectangle,
    sisRoundedRectangle, sisRoundedFilledRectangle);

  TSynIndicatorSpec = record
    Style: TSynIndicatorStyle;
    Foreground,
    Background: TD2D1ColorF;
    FontStyle: TFontStyles;
  end;

  TSynIndicator = record
    Id: TGuid;
    CharStart, CharEnd : Integer;
    constructor Create(aId: TGuid; aCharStart, aCharEnd: Integer);
  end;

  TSynIndicators = class
  public
    procedure RegisterSpec(Id: TGuid; Spec: TSynIndicatorSpec);
    function GetSpec(Id: TGUID): TSynIndicatorSpec;
    procedure Add(Line: Integer; Indicator: TSynIndicator);
    procedure Clear; overload;
    procedure Clear(Id: TGuid); overload;
    function LineIndicators(Line: Integer): TArray<TSynIndicator>;
    function IndicatorAtPos(Pos: TBufferCoord; var Indicator: TSynIndicator): Boolean;
    function IndicatorAtMousePos(MousePos: TPoint; var Indicator: TSynIndicator): Boolean;
```

SynEdit defines a public property

```pascal
property Indicators: TSynIndicators;
```

## Usage

You need to define and register an indicator spec, e.g.:

``` pascal
  const TestIndicatorID: TGUID  = '{A5D68F5B-7E52-4649-BD69-365B68B39A52}';

  IndicatorSpec: TSynIndicatorSpec =
    (Style: sisTextDecoration;
     Foreground: (r: 1; g:0; b: 0; a: 1);
     Background: (r: 0; g:0; b: 1; a: 0.5);
     FontStyle: [fsBold]);

 Editor.Indicators.RegisterSpec(TestIndicatorID, IndicatorSpec);
```

There are two options for adding/managing indicators:
### via a Plugin

The plugin needs to handle LinePut and LinesInserted to add indicators using for instance:
``` pascal
   Editor.Indicators.Add(aIndex, TSynIndicator.Create(TestIndicatorID, charStart, charEnd);
``` 

Note that SynEdit deletes indicators when their lines are deleted or edited and automatically adjusts the positions of other indicators.  Also SynEdit automatically invalidates indicators on removal.

### via an event handler

A new event  is defined that can directly provide indicators to the painting routine:

```pascal
type
  TGetLineIndicatorsEvent = procedure(Sender: TObject; const Line: Integer;
    var LineIndicators: TArray<TSynIndicator>) of Object;

  property OnGetLineIndicators: TGetLineIndicatorsEvent
    read FOnGetLineIndicators write FOnGetLineIndicators;
```

## Indicator styles

### sisTextDecoration
Can change style, font color and background color of the the indicated text.  The background can be solid or alpha blended:

![image](https://user-images.githubusercontent.com/1311616/155947533-7ec50bdb-1c36-4594-85c1-30dc1a9b486d.png)

### sisSquiggleMicrosoftWord

![image](https://user-images.githubusercontent.com/1311616/155947678-6febb92f-fe25-46d8-8d49-6c61048c509c.png)

### sisSquiggleWordPerfect 

![image](https://user-images.githubusercontent.com/1311616/155947949-bb4b1e4c-6fc1-4220-94fc-f6a87a815240.png)

### sisRectangle

![image](https://user-images.githubusercontent.com/1311616/155948121-50cd7963-540f-438c-bc5e-f05e882313c0.png)

### sisFilledRectangle

![image](https://user-images.githubusercontent.com/1311616/155948272-563195b3-7192-44d0-96bb-f917c3975d26.png)

### sisRoundedRectangle

![image](https://user-images.githubusercontent.com/1311616/155948362-b23e70d4-7cd6-4359-bcf4-f41ed11edcdc.png)

### sisRoundedFilledRectangle

![image](https://user-images.githubusercontent.com/1311616/155948484-bc5f2652-2d24-4c45-b00e-b0e45cc9a42b.png)

## Indicator hints and pop up menus

You can show an indicator specific hint or an an indicator specific popup by handling the OnShowHint and/or the OnContextPopup events and test for an indicator using IndicatorAtMousePos method.

## Bracket highlighting

You can use indicators instead of handling the PaintTransient event to implement bracket highlighting.  This is going to be a lot more efficient.  I am planning to implement bracket highlighting out-of-the-box, but more on this later.

## Demo

I will modify the spell checking component to use indicators and provide a demo.

@MShark67 @JaFi-cz @dados What do you think?

**Reply by JaFi-cz** [reply]
It seems nice, but maybe I am wrong.
E.g. on SpellCheck plugin works just now over visible area. You can add new words into dictionary, ignore words for check - it means content must be redraw. 
Similar is it with matching bracket/code parts (e.g. begin-end, HTML elements e.t.c.). Each code modification means level of matching brackets can be different and you need to delete indicator list and start from the begin.

**Reply by pyscripter** [reply]
Invalidation and painting when appropriate is handled automatically when indicators are added or removed.  Wait until I update the Spelling component.

**Reply by JaFi-cz** [reply]
I don't mean invalidation, I mean content of the indicators list.

**Reply by pyscripter** [reply]
The plugin is responsible for adding removing indicators, in the case of spelling, spelling errors.   It can do it by handling for instance the LinePut, LinesInserted events.   See section on usage.

**Reply by pyscripter** [reply]
@JaFi-cz By the way I am thinking of replacing the spell checking with the [following](https://en.delphipraxis.net/topic/2388-spell-checker-implementation/?tab=comments#comment-54646).   The main drawback is that you need Windows 8 or later.  But there are clear advantages...

**Reply by JaFi-cz** [reply]
I am little sceptical. In the past you were able to use Office spell check, but MS stops it. I am not sure, if he won't do the same in the future for this spellcheck or request fees for ussage.
And Just now I am slowly preparing last version supporting Windows XP. Windows 7 is still used, even they are not supported. So I won't stop Windows 7 support now.

Anyway, is MS spell check implementation fast enough to support on the fly check?

**Reply by pyscripter** [reply]
Just to clarify this is a built-in feature of Windows and has nothing to do with Microsoft Office.  You install dictionaries through Windows settings.  Windows takes care for storing words added to the dictionaries etc. And yes it is free, full featured and fast.  Look at the demo.

**Reply by pyscripter** [reply]
@JaFi-cz Modified SynSpellCheck.pas to use Indicators.  Have a look.

You can use the OnShowHint and OnContextPopUp to show special hints and popup menus for spelling errors.   I will do a demo of that.

**Comment by dados** [comment]
Out of the box bracket highlighting is just an excellent idea.
Indicators is something that I might look into using. Could be really nice and should make any program look more intelligent :)
I really don't have any opinion on spell checking as I cannot see myself using it.

I would be happy running some tests with a provided demo

**Comment by MShark67** [comment]
This looks really cool.  I did some tests just using the OnGetLineIndicators event and it worked well.  I hit an issue if you caret right such that the indicated string is partially shown as LeftChar progresses.  A range check error and some painting and caret movement issues.  Let me know if you need more steps to reproduce.

**Reply by pyscripter** [reply]
Reproduced. I see also some painting errors with some styles when scrolling left and right.   Will look into it.

**Reply by pyscripter** [reply]
Fixed

**Comment by ClementDoss** [comment]
Hi!
I would like to use indicators when searching for a text. I would like to highlight the words the user is searching. Do you have a sample code?

Thanks

**Reply by pyscripter** [reply]
Have a look at SynSpellCheck,pas which now uses indicators.  More is coming.

**Reply by Ilia-Nenashev-devX** [reply]
@ClementDoss, have you done this? Maybe by now you have a sample code for search results highlighting?

**Reply by pyscripter** [reply]
See https://github.com/TurboPack/SynEdit/blob/master/Demos/SearchReplaceDemo/uSearchHighlighter.pas

**Reply by Ilia-Nenashev-devX** [reply]
Is there any way to see the positions of all document indicators on the scrollbar?

**Reply by pyscripter** [reply]
Add a new custom scrollbar annotation and handle the OnGetInfo event.

**Reply by Ilia-Nenashev-devX** [reply]
Note: it seems does not works without using default style at `Project → Options → Application → Appearance`, which adds line like `TStyleManager.TrySetStyle('Windows10');` into the project file to enable styling hooks.

**Reply by pyscripter** [reply]
Annotated scrollbars only work with Vcl.Styles other than the system style.

**Reply by Ilia-Nenashev-devX** [reply]
Thanks!
Maybe indicator annotations could become a built-in indicator's feature in future releases?

```pascal
procedure AnnotateIndicators(
  ASynEdit: TSynEdit;
  const AIndicatorIdToAnnotate: TGUID; AIndicatorColorToAnnotate: TColor;
  var ARows: TArray<Integer>; var AColors: TArray<TColor>);
var
  LRowList: TList<Integer>;
  LLineIndicators: TArray<TSynIndicator>;
  I, J, K: Integer;
  LBufferCoord1, LBufferCoord2: TBufferCoord;
  LRow1, LRow2: Integer;
begin
  AColors := [AIndicatorColorToAnnotate];

  LRowList := TList<Integer>.Create;
  try
    for I := 0 to ASynEdit.Lines.Count - 1 do begin
      LLineIndicators := ASynEdit.Indicators.LineIndicators(I);

      for J := 0 to High(LLineIndicators) do begin
        if LLineIndicators[J].Id <> AIndicatorIdToAnnotate then
          Continue;

        LBufferCoord1 := BufferCoord(LLineIndicators[J].CharStart, I);
        LBufferCoord2 := BufferCoord(LLineIndicators[J].CharEnd, I);

        LRow1 := ASynEdit.BufferToDisplayPos(LBufferCoord1).Row;
        LRow2 := ASynEdit.BufferToDisplayPos(LBufferCoord2).Row;
        for K := LRow1 to LRow2 do
          LRowList.Add(K);
      end;
    end;
    ARows := LRowList.ToArray;
  finally
    LRowList.Free;
  end;
end;
```

And could You make `TCustomSynEdit.UpdateScrollBars` public to allow me force redraw changed annotations without direct using of `SendMessage(SynEdit1.Handle, WM_NCPAINT, 0, 0);`?

**Reply by pyscripter** [reply]
Please open new issues to make requests for new features.

The code you showed is quite inefficient.   Note also that Indicators do not cross lines. They start and stop on the same line.  

In any case in  https://github.com/pyscripter/SynEdit  I have made UpdateScrollBars public and added:

```pascal
function TSynIndicators.GetById(
  const Id: TGUID): TArray<TPair<Integer, TSynIndicator>>;
```
With this method your GetInfo could be as simple as:

```pascal
procedure TEditorForm.ScrollbarAnnotationGetInfo(Sender: TObject;
  AnnType: TSynScrollbarAnnType; var Rows: TArray<Integer>;
  var Colors: TArray<TColor>);
begin
  var Editor := Sender as TCustomSynEdit;
  Rows := [];
  if AnnType = sbaCustom1 then
  begin
    for var Pair in SynEdit.Indicators.GetById(YourIndicatorId) do
      Rows := Rows + [Editor.LineToRow(Pair.Key)];
    Colors := [YourColor];
  end;
  // Optinally sort and remove duplicates from Rows - but it is not needed
end;
```

**Reply by Ilia-Nenashev-devX** [reply]
ok, #272 registered.

But a single line can be wrapped over multiple rows, can't it?

For word-wrap you can use something like:

```pascal
procedure TEditorForm.ScrollbarAnnotationGetInfo(Sender: TObject;
  AnnType: TSynScrollbarAnnType; var Rows: TArray<Integer>;
  var Colors: TArray<TColor>);
begin
  var Editor := Sender as TCustomSynEdit;
  Rows := [];
  if AnnType = sbaCustom1 then
  begin
    for var Pair in SynEdit.Indicators.GetById(YourIndicatorId) do
    begin
      var BC := BufferCoord(Indicator.CharStart, Pair.Key);
      var DC := Editor.BufferToDisplayCorrd(BC);
      Rows := Rows + [DC.Row];
    end;
    Colors := [YourColor];
  end;
  // Optinally sort and remove duplicates from Rows - but it is not needed
end;
```

**Reply by Ilia-Nenashev-devX** [reply]
BTW, instead of TSynIndicators.GetById and every time make a partial copy of array, better to have an record-based skipping iterator.

**Reply by Ilia-Nenashev-devX** [reply]
And the code you included in my comment doesn't properly handle indicators that wrap into several rows — it annotates only the first one.

**Reply by pyscripter** [reply]
> And the code you included in my comment doesn't properly handle indicators that wrap into several rows — it annotates only the first one.

You seem to be a competent programmer.   How difficult is it to do the following, if you need to?:

```pascal
    for var Pair in SynEdit.Indicators.GetById(YourIndicatorId) do
    begin
      var BC := BufferCoord(Indicator.CharStart, Pair.Key);
      var DC := Editor.BufferToDisplayCorrd(BC);
      var StartRow := DC.Row;
      BC := BufferCoord(Indicator.CharStop, Pair.Key);
      DC := Editor.BufferToDisplayCorrd(BC);
      var EndRow := DC.Row;
      for var Row := StartRow to EndRow do
        Rows := Rows + [Row];
    end;
```

**Comment by MShark67** [comment]
Quick question!  I may be misunderstanding the CharStart and CharEnd values.  I was thinking that if I had a one character indicator, that started at position 1, that CharStart would be 1 and that CharEnd would be 1.   It seems like now it has to be CharStart 1 and CharEnd 2 for it to work.  Is that intended?

**Comment by pyscripter** [comment]
This was changed in the last commit.  It is now analogous to BlockBegin, BlockEnd used for selection.   So, CharEnd points to the character after the last character.

**Reply by MShark67** [reply]
Excellent, thanks!
